File: app/backend/hawkeye_backend/models/incident.py
"""Incidents, the call transcript, guidance instructions, and the sealed replay."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field

from hawkeye_backend.models.common import Provenance, utc_now
from hawkeye_backend.models.verification import VerificationResult


class IncidentType(StrEnum):
    """The two incident types. Nothing else classifies.

    Faint was the third until 2026-09-19. Fall detection was the weakest link
    in the chain - a debounce problem dressed as a clinical variable - and what
    a dispatcher actually needs is narrower and defensible: whether the people
    inside can respond. That is carried by respiration, per presence.
    """

    BURGLARY = "burglary"
    FIRE = "fire"


class RaisedBy(StrEnum):
    """Who raised it.

    USER is the only path to a call. Hawk Eye never dials 911 on its own;
    settled 2026-09-19. A human tap is what releases `agents/caller` to dial,
    and `agents/people` and `agents/master` surface their detections as
    interior state the resident acts on rather than as a call.

    SYSTEM is kept for wire compatibility and for records raised before that
    decision. It must never reach the dialing path: `assert_human_released` in
    `master/base.py` enforces that structurally rather than by convention.
    """

    USER = "user"
    SYSTEM = "system"


class IncidentStatus(StrEnum):
    RAISED = "raised"
    CLASSIFIED = "classified"
    CALLING = "calling"
    ON_CALL = "on_call"
    DISPATCHED = "dispatched"
    RESOLVED = "resolved"
    REFUSED = "refused"


class CallState(StrEnum):
    """State of the phone call to the 911 operator."""

    NOT_STARTED = "not_started"
    DIALING = "dialing"
    CONNECTED = "connected"
    ENDED = "ended"
    NOT_PLACED = "not_placed"


class IncidentClassification(BaseModel):
    """Why master called it what it called it.

    Classification is the interesting part and should be visible. Elevated CO
    plus a breathing signature that has gone missing is a fire with an occupant
    who may not be able to respond, and that is two independent modalities
    rather than one signal crossing a threshold.
    """

    incident_type: IncidentType
    reasoning: str = Field(description="Plain English, shown to the resident.")
    contributing_claim_ids: list[str] = Field(default_factory=list)
    discarded_claim_ids: list[str] = Field(
        default_factory=list,
        description="Claims that did not survive verification. Named, not hidden.",
    )
    confidence: float = Field(ge=0.0, le=1.0)


class ContextNote(BaseModel):
    """One line the resident typed into the "what is happening" box.

    Goes to master and becomes available to caller for the rest of the call.
    Provenance is `user-input`: it is a human statement, not a measurement, and
    caller must attribute it as such rather than asserting it as sensed fact.
    """

    note_id: str
    incident_id: str
    text: str = Field(min_length=1, max_length=2000)
    at: datetime = Field(default_factory=utc_now)
    provenance: Provenance
    delivered_to_caller: bool = False


class TranscriptSpeaker(StrEnum):
    """Who said a line.

    CALLER is our agent's synthesized voice. OPERATOR is the human dispatcher.
    RESIDENT appears when the resident speaks on the line directly.
    SYSTEM is a non-speech annotation, e.g. "call connected".
    """

    CALLER = "caller"
    OPERATOR = "operator"
    RESIDENT = "resident"
    SYSTEM = "system"


class TranscriptLine(BaseModel):
    """One line of the caller <-> 911 conversation, shown to the resident.

    The operator does not see this. It exists so the resident is not left in
    silence while a synthetic voice speaks on their behalf.
    """

    line_id: str
    incident_id: str
    speaker: TranscriptSpeaker
    text: str
    at: datetime = Field(default_factory=utc_now)
    final: bool = Field(
        default=True,
        description="False for a partial transcription being revised in place. Match on line_id.",
    )
    claim_ids: list[str] = Field(
        default_factory=list,
        description="Verified claims this line repeats. Lets the app link a spoken sentence to its proof.",
    )
    provenance: Provenance


class InstructionOrigin(StrEnum):
    """Where an instruction came from.

    The UI should not distinguish these, because the user does not care. The API
    does, because the safety rules differ: relayed operator instructions always
    win over generated first aid.
    """

    RELAYED_OPERATOR = "relayed_operator"
    FIRST_AID = "first_aid"
    SYSTEM_STATUS = "system_status"


class Instruction(BaseModel):
    """One thing agents/caller is telling the resident to do."""

    instruction_id: str
    incident_id: str
    text: str
    origin: InstructionOrigin
    at: datetime = Field(default_factory=utc_now)
    urgent: bool = False
    supersedes_instruction_id: str | None = None
    defers_to_operator: bool = Field(
        default=True,
        description=(
            "True when this instruction must be dropped the moment the dispatcher gives a "
            "competing one. Dispatchers are trained for this and the agent is not."
        ),
    )
    provenance: Provenance


class Incident(BaseModel):
    """One incident, from raised to resolved."""

    incident_id: str
    site_id: str
    incident_type: IncidentType
    status: IncidentStatus
    raised_by: RaisedBy
    raised_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)
    resolved_at: datetime | None = None
    address: str = Field(description="What the caller will read to the dispatcher.")
    classification: IncidentClassification | None = None
    call_state: CallState = CallState.NOT_STARTED
    context_notes: list[ContextNote] = Field(default_factory=list)
    refusal_reason: str | None = Field(
        default=None,
        description="Set when status is REFUSED: why the system declined to escalate.",
    )


class RaiseIncidentRequest(BaseModel):
    """POST /v1/incident body."""

    incident_type: IncidentType
    note: str | None = Field(
        default=None,
        max_length=2000,
        description="Optional first line of context, same channel as the 'what is happening' box.",
    )


class IncidentAck(BaseModel):
    """POST /v1/incident response. Deliberately small; the stream carries the rest."""

    incident_id: str
    status: IncidentStatus
    accepted_at: datetime = Field(default_factory=utc_now)


class ContextRequest(BaseModel):
    """POST /v1/incident/{id}/context body."""

    text: str = Field(min_length=1, max_length=2000)
    speak_on_call: bool = Field(
        default=False,
        description=(
            "Also route this note to agents/caller so it is spoken on an "
            "already-running call, attributed to the resident. Best-effort: "
            "a failure to speak it must never fail storing the note itself. "
            "Context, never instruction - it cannot change what caller "
            "trusts or where the incident is directed."
        ),
    )


class ReplayEntry(BaseModel):
    """One sealed entry in the post-incident record."""

    seq: int
    at: datetime
    kind: str = Field(
        description=(
            "lifecycle | frame | incident | transcript | instruction | verification | "
            "context | notice | state"
        )
    )
    actor: str | None = Field(
        default=None,
        description=(
            "Who is responsible for this entry: an ANSName, 'hub', '911-operator', "
            "'resident'. Null on entries written before the recorder tracked it, which "
            "is why it is optional rather than required."
        ),
    )
    summary: str
    detail: dict[str, object] = Field(default_factory=dict)
    entry_hash: str = Field(description="SHA-256 over the canonical JSON of this entry.")
    prev_hash: str | None = Field(
        default=None, description="Hash chain. Ties this entry to the one before it."
    )


class ReplayRecord(BaseModel):
    """GET /v1/incident/{id}/replay response, from agents/replay.

    Two audiences: detectives after a burglary, and accountability for the system
    itself. The operator cannot verify us live; an investigator can verify this
    afterward, and swatting investigations are entirely post-hoc.
    """

    incident_id: str
    sealed: bool
    sealed_at: datetime | None = None
    caller_ansname: str
    site_address: str
    entries: list[ReplayEntry]
    verifications: list[VerificationResult] = Field(
        default_factory=list, description="Every verification, accepted and discarded alike."
    )
    root_hash: str | None = Field(
        default=None, description="Hash of the final entry. What gets submitted to SCITT."
    )
    scitt_receipt: str | None = Field(
        default=None,
        description="SCITT transparency log receipt. Null until the entry is sealed.",
    )
    # TODO(ans): the SCITT submission and receipt format is not implemented here.
    # Question to answer against https://github.com/agentnameservice/ans-registry:
    # what is the exact submit endpoint and receipt structure for an ANS SCITT
    # transparency log entry (ANS-4), and is a receipt a COSE object or a JSON
    # document? Until that is answered this field stays null and the README says
    # so. Do not fabricate a receipt: an unverifiable receipt is worse than none.
    hash_algorithm: str = "sha256"

File: app/ios/HawkEye/Models/Incident.swift
import Foundation
import SwiftUI

/// The two incident types. Fixed set, per the root `CLAUDE.md`.
enum IncidentType: String, Codable, Sendable, Hashable, CaseIterable, Identifiable {
    case burglary, fire

    var id: String { rawValue }

    var title: String {
        switch self {
        case .burglary: "Burglary"
        case .fire: "Fire"
        }
    }

    var symbol: String {
        switch self {
        case .burglary: "figure.run"
        case .fire: "flame.fill"
        }
    }

    var tint: Color {
        switch self {
        case .burglary: Palette.burglary
        case .fire: Palette.fire
        }
    }
}

extension IncidentType {

    /// The one incident type the pivot left.
    ///
    /// Fire went with the simulated gas sensor on 2026-09-19 and fall detection
    /// went earlier the same day, so there is nothing for a resident to choose,
    /// which is the right shape for a control someone uses while frightened.
    ///
    /// The wire enum still spells it `burglary` until T01 renames it. Every
    /// caller that raises an incident goes through this one name instead, so the
    /// rename is a one-line change here rather than a search across the app, and
    /// nothing user-facing says the old word.
    static var intrusion: IncidentType { .burglary }
}

/// How the incident was raised. The autonomous path is the one that matters,
/// because the person who would have pressed the button may not be in a state
/// to press it.
enum IncidentOrigin: String, Codable, Sendable, Hashable, CaseIterable {
    case user
    case system

    var label: String {
        switch self {
        case .user: "Raised by you"
        case .system: "Raised by Hawk Eye"
        }
    }
}

/// Where the incident is in its lifecycle, as `master` reports it.
enum IncidentStatus: String, Codable, Sendable, Hashable, CaseIterable {
    case raised, classified, calling
    case onCall = "on_call"
    case dispatched, resolved, refused
}

/// Which transition the `incident` event is reporting.
enum IncidentPhase: String, Codable, Sendable, Hashable, CaseIterable {
    case raised, classified, updated, resolved, refused
}

/// State of the phone call to the 911 operator.
enum CallState: String, Codable, Sendable, Hashable, CaseIterable {
    case notStarted = "not_started"
    case dialing
    case connected
    case ended
    case notPlaced = "not_placed"

    var label: String {
        switch self {
        case .notStarted: "Preparing to call"
        case .dialing: "Dialing 911"
        case .connected: "On the line with 911"
        case .ended: "Call ended"
        case .notPlaced: "No call placed"
        }
    }
}

/// Why `master` called it what it called it.
///
/// Classification is the interesting part and should be visible.
/// Elevated CO plus a breathing signature that has gone missing is a fire with
/// an occupant who may not be able to respond, not two separate incidents.
struct IncidentClassification: Codable, Sendable, Hashable {
    var incidentType: IncidentType
    /// Plain English, shown to the resident.
    var reasoning: String
    var contributingClaimIDs: [String] = []
    /// Claims that did not survive verification. Named, not hidden.
    var discardedClaimIDs: [String] = []
    var confidence: Double = 0

    enum CodingKeys: String, CodingKey {
        case reasoning, confidence
        case incidentType = "incident_type"
        case contributingClaimIDs = "contributing_claim_ids"
        case discardedClaimIDs = "discarded_claim_ids"
    }
}

/// One line the resident typed into the "what is happening" box, echoed back so
/// the app can confirm delivery.
///
/// Provenance is `user-input`: it is a human statement, not a measurement, and
/// `caller` attributes it as one rather than asserting it as sensed fact.
struct ContextNote: Codable, Sendable, Hashable, Identifiable {
    var noteID: String
    var incidentID: String
    var text: String
    var at: Date = .distantPast
    var provenance: Provenance
    var deliveredToCaller: Bool = false

    var id: String { noteID }

    enum CodingKeys: String, CodingKey {
        case text, at, provenance
        case noteID = "note_id"
        case incidentID = "incident_id"
        case deliveredToCaller = "delivered_to_caller"
    }
}

/// One incident, from raised to resolved. The backend's `Incident`, field for
/// field.
struct Incident: Codable, Sendable, Hashable, Identifiable {
    var incidentID: String
    var siteID: String
    var incidentType: IncidentType
    var status: IncidentStatus
    var raisedBy: IncidentOrigin
    var raisedAt: Date = .distantPast
    var updatedAt: Date = .distantPast
    var resolvedAt: Date?
    /// What `caller` will read to the dispatcher.
    var address: String = ""
    var classification: IncidentClassification?
    var callState: CallState = .notStarted
    var contextNotes: [ContextNote] = []
    /// Set when status is `refused`: why the system declined to escalate.
    var refusalReason: String?

    var id: String { incidentID }
    var type: IncidentType { incidentType }
    var origin: IncidentOrigin { raisedBy }
    var openedAt: Date { raisedAt }
    /// `master`'s reasoning, in plain English, or the refusal when it declined.
    var summary: String? { refusalReason ?? classification?.reasoning }

    enum CodingKeys: String, CodingKey {
        case status, address, classification
        case incidentID = "incident_id"
        case siteID = "site_id"
        case incidentType = "incident_type"
        case raisedBy = "raised_by"
        case raisedAt = "raised_at"
        case updatedAt = "updated_at"
        case resolvedAt = "resolved_at"
        case callState = "call_state"
        case contextNotes = "context_notes"
        case refusalReason = "refusal_reason"
    }

    init(
        incidentID: String,
        siteID: String,
        incidentType: IncidentType,
        status: IncidentStatus,
        raisedBy: IncidentOrigin,
        raisedAt: Date,
        updatedAt: Date,
        resolvedAt: Date? = nil,
        address: String,
        classification: IncidentClassification? = nil,
        callState: CallState = .notStarted,
        contextNotes: [ContextNote] = [],
        refusalReason: String? = nil
    ) {
        self.incidentID = incidentID
        self.siteID = siteID
        self.incidentType = incidentType
        self.status = status
        self.raisedBy = raisedBy
        self.raisedAt = raisedAt
        self.updatedAt = updatedAt
        self.resolvedAt = resolvedAt
        self.address = address
        self.classification = classification
        self.callState = callState
        self.contextNotes = contextNotes
        self.refusalReason = refusalReason
    }

    init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        incidentID = try c.decode(String.self, forKey: .incidentID)
        siteID = try c.decodeIfPresent(String.self, forKey: .siteID) ?? ""
        incidentType = try c.decode(IncidentType.self, forKey: .incidentType)
        status = try c.decode(IncidentStatus.self, forKey: .status)
        raisedBy = try c.decode(IncidentOrigin.self, forKey: .raisedBy)
        raisedAt = try c.decodeIfPresent(Date.self, forKey: .raisedAt) ?? Date()
        updatedAt = try c.decodeIfPresent(Date.self, forKey: .updatedAt) ?? raisedAt
        resolvedAt = try c.decodeIfPresent(Date.self, forKey: .resolvedAt)
        address = try c.decodeIfPresent(String.self, forKey: .address) ?? ""
        classification = try c.decodeIfPresent(IncidentClassification.self, forKey: .classification)
        callState = try c.decodeIfPresent(CallState.self, forKey: .callState) ?? .notStarted
        contextNotes = try c.decodeIfPresent([ContextNote].self, forKey: .contextNotes) ?? []
        refusalReason = try c.decodeIfPresent(String.self, forKey: .refusalReason)
    }
}

/// `POST /v1/incident` request and its 202 acknowledgement.
struct RaiseIncidentRequest: Codable, Sendable, Hashable {
    var incidentType: IncidentType
    var note: String?

    enum CodingKeys: String, CodingKey {
        case note
        case incidentType = "incident_type"
    }
}

struct IncidentAck: Codable, Sendable, Hashable {
    var incidentID: String
    var status: IncidentStatus
    var acceptedAt: Date = .distantPast

    enum CodingKeys: String, CodingKey {
        case status
        case incidentID = "incident_id"
        case acceptedAt = "accepted_at"
    }
}

/// `POST /v1/incident/{id}/context` request body.
struct ContextRequest: Codable, Sendable, Hashable {
    var text: String
}

/// One line of the caller to 911 conversation, shown to the resident.
///
/// The operator does not see this. It exists so the resident is not left in
/// silence while a synthetic voice speaks on their behalf.
struct TranscriptLine: Codable, Sendable, Hashable, Identifiable {

    enum Speaker: String, Codable, Sendable, Hashable, CaseIterable {
        /// `agents/caller`, speaking on the resident's behalf.
        case caller
        /// The human 911 operator.
        case operatorVoice = "operator"
        /// The resident, speaking on the line directly.
        case resident
        /// A non-speech annotation, e.g. "call connected".
        case system

        var label: String {
            switch self {
            case .caller: "Hawk Eye"
            case .operatorVoice: "911 Operator"
            case .resident: "You"
            case .system: "Call"
            }
        }
    }

    var lineID: String
    var incidentID: String
    var speaker: Speaker
    var text: String
    var at: Date = .distantPast
    /// False for a partial transcription being revised in place.
    var final: Bool = true
    /// Verified claims this line repeats, so a spoken sentence can be linked to
    /// its proof.
    var claimIDs: [String] = []
    var provenance: Provenance?

    var id: String { lineID }
    /// The view reads this. A line still being revised is drawn dimmer.
    var partial: Bool { !final }

    enum CodingKeys: String, CodingKey {
        case speaker, text, at, final, provenance
        case lineID = "line_id"
        case incidentID = "incident_id"
        case claimIDs = "claim_ids"
    }

    init(
        lineID: String,
        incidentID: String,
        speaker: Speaker,
        text: String,
        at: Date,
        final: Bool = true,
        claimIDs: [String] = [],
        provenance: Provenance? = nil
    ) {
        self.lineID = lineID
        self.incidentID = incidentID
        self.speaker = speaker
        self.text = text
        self.at = at
        self.final = final
        self.claimIDs = claimIDs
        self.provenance = provenance
    }

    init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        lineID = try c.decode(String.self, forKey: .lineID)
        incidentID = try c.decodeIfPresent(String.self, forKey: .incidentID) ?? ""
        speaker = try c.decode(Speaker.self, forKey: .speaker)
        text = try c.decode(String.self, forKey: .text)
        at = try c.decodeIfPresent(Date.self, forKey: .at) ?? Date()
        final = try c.decodeIfPresent(Bool.self, forKey: .final) ?? true
        claimIDs = try c.decodeIfPresent([String].self, forKey: .claimIDs) ?? []
        provenance = try c.decodeIfPresent(Provenance.self, forKey: .provenance)
    }
}

/// Where an instruction came from.
///
/// The UI deliberately does not distinguish these, because the user does not
/// care. The API does, because the safety rules differ: relayed operator
/// instructions always win over generated first aid.
enum InstructionOrigin: String, Codable, Sendable, Hashable, CaseIterable {
    case relayedOperator = "relayed_operator"
    case firstAid = "first_aid"
    case systemStatus = "system_status"
}

/// One thing `agents/caller` is telling the resident to do.
///
/// **The client never authors this text.** Everything shown comes from the
/// guidance agent, which is the component that owns the safety rules in
/// `agents/CLAUDE.md`. Hardcoding first-aid copy in the client would put
/// medical instructions outside the one place reviewed for them.
struct Instruction: Codable, Sendable, Hashable, Identifiable {
    var instructionID: String
    var incidentID: String
    var text: String
    var origin: InstructionOrigin
    var at: Date = .distantPast
    /// Guidance marks the small number of instructions that are time-critical.
    /// This changes weight and nothing else; it never reveals the source.
    var urgent: Bool = false
    var supersedesInstructionID: String?
    /// True when this instruction must be dropped the moment the dispatcher
    /// gives a competing one. Dispatchers are trained for this; the agent is not.
    var defersToOperator: Bool = true
    var provenance: Provenance?

    var id: String { instructionID }

    enum CodingKeys: String, CodingKey {
        case text, origin, at, urgent, provenance
        case instructionID = "instruction_id"
        case incidentID = "incident_id"
        case supersedesInstructionID = "supersedes_instruction_id"
        case defersToOperator = "defers_to_operator"
    }

    init(
        instructionID: String,
        incidentID: String,
        text: String,
        origin: InstructionOrigin,
        at: Date,
        urgent: Bool = false,
        supersedesInstructionID: String? = nil,
        defersToOperator: Bool = true,
        provenance: Provenance? = nil
    ) {
        self.instructionID = instructionID
        self.incidentID = incidentID
        self.text = text
        self.origin = origin
        self.at = at
        self.urgent = urgent
        self.supersedesInstructionID = supersedesInstructionID
        self.defersToOperator = defersToOperator
        self.provenance = provenance
    }

    init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        instructionID = try c.decode(String.self, forKey: .instructionID)
        incidentID = try c.decodeIfPresent(String.self, forKey: .incidentID) ?? ""
        text = try c.decode(String.self, forKey: .text)
        origin = try c.decode(InstructionOrigin.self, forKey: .origin)
        at = try c.decodeIfPresent(Date.self, forKey: .at) ?? Date()
        urgent = try c.decodeIfPresent(Bool.self, forKey: .urgent) ?? false
        supersedesInstructionID = try c.decodeIfPresent(String.self, forKey: .supersedesInstructionID)
        defersToOperator = try c.decodeIfPresent(Bool.self, forKey: .defersToOperator) ?? true
        provenance = try c.decodeIfPresent(Provenance.self, forKey: .provenance)
    }
}
